src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
+  }
+  AstExpr(const AstExpr& other) {
+    ptr = move(other.ptr);
```

Comment:
I don't think we're supposed to move stuff from `const` objects :)
But I think this actually doesn't call `ptr`'s move assignment as (I think?) that `move(other.ptr)` here would return a `const&&`, which is a weird thing that's not useful

Reply:
Anyway, I'd just default this (`AstExpr(const AstExpr& other) = default`), it's simpler and less error prone (also future proof)

Reply:
This only works because `ptr` is mutable :scream: 

Copying `unique_ptr`s to virtual objects is problematic. You usually want a virtual `copy()` method and to call it manually in copy constructor/assignment if you need copying.

(`shared_ptr` is indeed much nicer)